src/main/java/org/spongepowered/api/item/recipe/crafting/ShapedCraftingRecipe.java

```diff
@@ -0,0 +1,161 @@
+     */
+    int getHeight();
+
+    interface Builder extends CraftingRecipe.Builder<ShapedCraftingRecipe, Builder> {
+
+        // TODO use #aisle and #where or....
+
+        /**
+         * Sets the aisle pattern for the shaped recipe and resets the
+         * registered ingredient predicates.
+         * <p>An aisle represents a single row of the crafting inventory
+         * input.
+         * Use the {@link #where(char, Ingredient)} to assign ingredient
+         * predicates to characters of the aisles.
+         * Setting the aisle resets all previously assigned {@link Ingredient}s.
+         * The space character will be defaulted to {@link Ingredient#NONE} if not specified.
+         * Any other not assigned characters will cause an Exception when {@link #build()}ing the Recipe.
+         *
+         * TODO Example
+         *
+         * @param aisle A string array of ingredients
+         * @return The builder
+         */
+        Builder aisle(String... aisle);
+
+        /**
+         * Sets an ingredient predicate based on the aisle pattern for the
+         * shaped recipe. Sets the ingredient to {@link Ingredient#NONE} if {@code null} is
+         * specified.
+         *
+         * @param symbol The ingredient symbol
+         * @param ingredient The ingredient predicate to set
+         * @return The builder
+         * @throws IllegalArgumentException If the aisle does not contain
+         *     the specified character symbol
+         */
+        Builder where(char symbol, @Nullable Ingredient ingredient) throws IllegalArgumentException;
+
+        // TODO ... or use #row
+
+        /**
+         * Adds a row of ingredients.
+         *
+         * @param ingredients The row of ingredients.
+         * @return This builder
+         */
+        default Builder row(Ingredient... ingredients) {
+            return this.row(0, ingredients);
+        }
+
+        /**
+         * Adds a row of ingredients filling the skipped columns with {@link Ingredient#NONE}
+         *
+         * @param skip The amount of columns to skip.
+         * @param ingredients The row of ingredients.
+         * @return This builder
+         */
+        Builder row(int skip, Ingredient... ingredients);
+
+        // ------------------------------------------------
+
+        /**
+         * Sets the resultant {@link ItemStackSnapshot} for when this shaped
+         * recipe is correctly crafted.
+         *
+         * @param result The resultant snapshot
+         * @return The builder
+         */
+        default Builder result(ItemStackSnapshot result) {
+            checkNotNull(result, "result");
+            return this.result(result.createStack());
+        }
+
+        /**
+         * Sets the resultant {@link ItemStack} for when this shaped recipe
+         * is correctly crafted.
+         *
+         * @param result The resultant stack
+         * @return The builder
+         */
+        Builder result(ItemStack result);
```

Comment:
As long as the item stack don't change after they are sent into the registry, things will be fine. We just need to copy it in the implementation before we send it into the registry.